### PR TITLE
Vary initial nonce in mock Praos and pure Shelley ThreadNet tests

### LINKED ISSUE #2235: Randomize initial nonce in Praos tests
In mock `Praos` and `RealTPraos`, the genesis block is currently initialized with the same fixed nonce every time. If I understand correctly, that means the leader schedule for the first epoch is similarly fixed. Especially once we start with statistical testing (eg for CG/CP violation rates), we'll want our samples to be more independent than that.

Thus this Issue is to have the tests' QuickCheck generators determine the initial nonce.

### PR DESCRIPTION
**Pull request by nfrisby**
Fixes #2235.

### CODE REVIEW

**Comment by nfrisby** [comment]
@mrBliss good catch. I had that diff stranded on a separate WIP branch.

I've copied it over here, rebased onto latest `master`, and now force-pushed. So this PR now handles mock Praos and pure Shelley (ie `RealTPraos`) ThreadNet tests.

P.S. My upcoming PR for Issue #2388 (which your Approval on PR 2490 has unblocked; thanks!) will make a similar change to the Cardano ThreadNet tests.

File: ouroboros-consensus-shelley-test/test/Test/ThreadNet/RealTPraos.hs [hunk 1]
```diff
@@ -78,6 +83,11 @@ instance Arbitrary TestSetup where
       setupD  <- arbitrary
       setupD2 <- arbitrary
 
+      setupInitialNonce <-
```
Comment (nfrisby) [inline ouroboros-consensus-shelley-test/test/Test/ThreadNet/RealTPraos.hs:86]: Is there a preferable path for this hash generator?

File: ouroboros-consensus/ouroboros-consensus-mock/test/Test/ThreadNet/Praos.hs [hunk 2]
```diff
@@ -122,5 +122,6 @@ prop_simple_praos_convergence TestSetup
                                       (HardFork.defaultEraParams
                                         k
                                         slotLength)
+                                      0
```
Comment (nfrisby) [inline ouroboros-consensus/ouroboros-consensus-mock/test/Test/ThreadNet/Praos.hs:125]: Hmmm, I must have lost part of the diff. Will fix this degenerate argument shortly.

File: ouroboros-consensus-shelley-test/test/Test/ThreadNet/RealTPraos.hs (hunk 1 above)
Comment (mrBliss) [reply ouroboros-consensus-shelley-test/test/Test/ThreadNet/RealTPraos.hs:86]: Yeah, I'd go for [`mkNonceFromNumber`](https://github.com/input-output-hk/cardano-ledger-specs/blob/ed7acfc16efba51e10bad3e8fc7f19b043fd8a9a/shelley/chain-and-ledger/executable-spec/src/Shelley/Spec/Ledger/BaseTypes.hs#L201). Maybe we should also sometimes generate `NeutralNonce`.

Comment (nfrisby) [reply ouroboros-consensus-shelley-test/test/Test/ThreadNet/RealTPraos.hs]: Done, thanks!

**Comment by nfrisby** [comment]
I fixed the last 2 mini tasks and optimistically rebased onto bors' latest candidate. I think in a a few tens of minutes this PR will be back down to two commits vs `master` and ready for Approval.

File: ouroboros-consensus-shelley-test/test/Test/ThreadNet/RealTPraos.hs [hunk 3]
```diff
@@ -78,6 +82,11 @@ instance Arbitrary TestSetup where
       setupD  <- arbitrary
       setupD2 <- arbitrary
 
+      setupInitialNonce <- frequency
+        [ (1, pure SL.NeutralNonce)
+        , (9, SL.mkNonceFromNumber <$> arbitrary)
+        ]
```
Comment (mrBliss) [inline ouroboros-consensus-shelley-test/test/Test/ThreadNet/RealTPraos.hs:88]: Exactly what I had in mind :slightly_smiling_face:

**Comment by mrBliss** [comment]
> I fixed the last 2 mini tasks and optimistically rebased onto bors' latest candidate. I think in a a few tens of minutes this PR will be back down to two commits vs `master` and ready for Approval.

Perfect!

**Comment by nfrisby** [comment]
I `touch`ed the last commit and then force pushed, because GitHub was confused. Its UI was listing that last few commits merged into `master` as part of this PR. This is almost certainly because I had rebased onto bors' staging commit. I won't do that again :(

**Comment by mrBliss** [comment]
> I `touch`ed the last commit and then force pushed, because GitHub was confused. Its UI was listing the last few commits merged into `master` as part of this PR. This is almost certainly because I had rebased onto bors' staging commit. I won't do that again :(

I have done that a few times in the past and as far as I can remember, it always worked. Maybe just bad luck this time.

**Pull request merged**